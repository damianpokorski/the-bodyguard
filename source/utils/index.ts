export * from './logging';
export * from './strings';
export * from './utils';
