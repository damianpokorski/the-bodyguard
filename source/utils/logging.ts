import { default as cliColor } from 'cli-color';
export const log = (message?: string, newline = true) =>
  process.stdout.write(`${message ?? ''}${newline ? '\n' : ''}`);

export const listItem = (item: string) => log(` - ${item}`);
export const ln = () => log('');

export const success = (message?: string, newline = true) => {
  log(`  ${cliColor.greenBright(`✔`)} ${message ?? ''}`, newline);
};

export const error = (message?: string, newline = true) =>
  log(`  ${cliColor.red(`✖`)} ${message ?? ''}`, newline);
export const warn = (message?: string, newline = true) =>
  log(`  ${cliColor.yellow(`!`)} ${message ?? ''}`, newline);

export const abort = (message: string, newline = true) => {
  error(message, newline);
  process.exit(1);
};

export const rollbackLine = () => {
  process?.stdout?.clearLine(0);
  process?.stdout?.cursorTo(0);
};
