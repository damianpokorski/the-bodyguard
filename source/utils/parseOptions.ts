import { join } from 'path/posix';

// Base options
export interface Options {
  openapi: string;
  output: string;
  // flags
  force?: boolean;
  sourcemaps?: boolean;
}

// Generated options
export interface InferredOptions extends Options {
  paths: {
    main: string;
    barrel: string;
    barrelDeclarations: string;
    models: string;
    schemas: string;
    tmp: string;
    validators: string;
    dist: string;
  };
  checksums: string;
}
export const parseOptions = (input: Options): InferredOptions => ({
  ...input,
  paths: {
    main: input.output,
    barrel: join(process.cwd(), input.output, 'index.ts'),
    barrelDeclarations: join(process.cwd(), input.output, 'index.d.ts'),
    models: join(process.cwd(), input.output, 'models'),
    schemas: join(process.cwd(), input.output, 'schemas'),
    validators: join(process.cwd(), input.output, 'validators'),
    dist: join(process.cwd(), input.output, 'dist'),
    tmp: join(process.cwd(), input.output, 'tmp')
  },
  checksums: join(process.cwd(), input.output, 'dist', 'checksums.json')
});
