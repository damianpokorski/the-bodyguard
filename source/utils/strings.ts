export const exceptions = {
    missingPackages: `Some of the required packages is missing`,
    openApiFileDoesNotExist: `File does not exists`,
    invalidOpenApiFile: `Failed to parse OpenAPI Specification`,
    failedToGenerateModelsUsingOpenAPIGenerator: "Failed to trigger OpenAPI-generator-cli",
    failedToGenerateModels: "Failed to generate models",
    failedToGenerateBundleDeclarations: "Failed to generate declarations using Typescript",
    failedToGenerateBundleEsbuild: "Failed to generate a minified bundle (esbuild)",
    failedToGenerateBundle: "Failed to generate bundle",
    failedToGenerateValidators: "Failed to generate descriptors using AJV & better-ajv-errors"
}

export const progress = {

}