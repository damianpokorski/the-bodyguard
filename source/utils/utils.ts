import { default as cliColor } from 'cli-color';
import { spawnSync } from 'node:child_process';
import { join } from 'path';

export const log = (message?: string, newline = true) => process.stdout.write(`${(message ?? '')}${newline ? '\n' : ''}`);
export const listItem = (item: string) => log(` - ${item}`);
export const ln = () => log('');
export const success = (message?: string, newline = true) => {
    log(`  ${cliColor.greenBright(`✔`)} ${message ?? ''}`, newline);
}
export const error = (message?: string, newline = true) => log(`  ${cliColor.red(`✖`)} ${message ?? ''}`, newline);
export const warn = (message?: string, newline = true) => log(`  ${cliColor.yellow(`!`)} ${message ?? ''}`, newline);
export const abort = (message: string, newline = true) => {
    error(message, newline);
    process.exit(1);
}

export const rollbackLine = () => {
    process?.stdout?.clearLine(0);
    process?.stdout?.cursorTo(0);
}

export interface Options {
    openapi: string;
    output: string;
};

export interface InferredOptions extends Options {
    paths: {
        main: string;
        barrel: string;
        barrelDeclarations: string;
        models: string;
        schemas: string;
        tmp: string;
        validators: string;
        dist: string;
    };
}

export const parseOptions = (input: Options): InferredOptions => ({
    ...input,
    paths: {
        main: input.output,
        barrel: join(process.cwd(), input.output, 'index.ts'),
        barrelDeclarations: join(process.cwd(), input.output, 'index.d.ts'),
        models: join(process.cwd(), input.output, 'models'),
        schemas: join(process.cwd(), input.output, 'schemas'),
        validators: join(process.cwd(), input.output, 'validators'),
        dist: join(process.cwd(), input.output, 'dist'),
        tmp: join(process.cwd(), input.output, 'tmp')
    }
});

export const defaultEncoding = { encoding: 'utf-8' as BufferEncoding };

export class SafetyNetBuildException extends Error {
    constructor(msg?: string, private innerError?: Error | unknown) {
        super(
            [
                msg,
                ((innerError ?? {}) as unknown as Error)['message'] ?? undefined
            ].filter(isValid => isValid).join(' - ')
        );
    }
}

export const exec = async (cmd: string, ...args: string[]): Promise<[boolean, string, string]> => {
    const result = spawnSync(cmd, args, {});
    return [
        result.status == 0 && result.error == undefined,
        result.stdout.toString(),
        result.stderr.toString()
    ];
}

export const npxPackageAvailable = async (packageName: string) => {
    // Try local package search first
    let [success, stdout, stderr] = (await exec(`npm`, 'ls', packageName, '--depth=0'));
    if (!success) {
        // Try global packages next
        [success, stdout, stderr] = (await exec(`npm`, 'ls', packageName, '--global'));
    }
    return success;
}
