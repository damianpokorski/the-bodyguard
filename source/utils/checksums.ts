import { existsSync, readFileSync, writeFileSync } from 'fs';
import * as md5 from 'md5-file';
import { join } from 'path';
import { log } from './logging';
import { InferredOptions } from './utils';

export const fileChecksum = (path: string) => {
  if (existsSync(path)) {
    return md5.sync(path);
  }
  return null;
};

export const generateChecksumsFile = (opts: InferredOptions) => {
  writeFileSync(
    opts.checksums,
    JSON.stringify(
      [
        opts.openapi,
        join(opts.paths.dist, `index.js`),
        join(opts.paths.dist, `index.d.ts`)
      ]
        .map((path) => ({ [path]: fileChecksum(path) }))
        .reduce((pathEntry, paths) => ({ ...pathEntry, ...paths }))
    )
  );
};

export const validateChecksums = (opts: InferredOptions) => {
  log(`Looking for past checksums..`);
  if (opts.force || !existsSync(opts.checksums)) {
    return false;
  }
  const checksums = JSON.parse(
    readFileSync(opts.checksums, { encoding: 'utf-8' })
  ) as Record<string, string | null>;

  for (const [file, pastChecksum] of Object.entries(checksums)) {
    const newChecksum = fileChecksum(file);
    if (newChecksum !== pastChecksum) {
      return false;
    }
  }

  log(
    `Checksums match, skipping generation... use --force if you want to regenerate anyways`
  );
  return true;
};
