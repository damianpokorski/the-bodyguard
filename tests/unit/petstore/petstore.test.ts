import { main } from '@main';
import { exceptions } from '@utils/strings';
import { Options } from '@utils/utils';
import { cleanUpDir, exceptionBuilder } from '../unit.utils';

const testDirs = [
  './tests/.builds.invalid-file-path',
  './tests/.builds.sample-generation'
];

describe('Petstore generation tests', () => {
  jest.setTimeout(60000);
  jest.useFakeTimers();

  // Prearrange
  process.stdout.clearLine = process.stdout.clearLine ?? jest.fn();
  process.stdout.cursorTo = process.stdout.cursorTo ?? jest.fn();
  const mocks = {
    clearLine: jest
      .spyOn(process.stdout, 'clearLine')
      .mockImplementation(jest.fn()),
    cursorTo: jest
      .spyOn(process.stdout, 'cursorTo')
      .mockImplementation(jest.fn()),
    write: jest.spyOn(process.stdout, 'write').mockImplementation(jest.fn())
  };

  // Cleaning up temporarary dirs, before and after each run
  const cleanUp = () => {
    for (const testDir of testDirs) {
      cleanUpDir(testDir);
    }
  };
  beforeEach(() => cleanUp());
  afterEach(() => cleanUp());

  describe('Fails to generate', () => {
    it('Fails to generate with missing OpenAPI', async () => {
      // Arrange
      const opts: Options = {
        openapi: './invalid-file.yaml',
        output: testDirs[0]
      };

      // Act
      const result = main(opts);

      // Assert
      expect(result).rejects.toThrow(
        exceptionBuilder([
          exceptions.invalidOpenApiFile,
          exceptions.openApiFileDoesNotExist,
          opts.openapi
        ])
      );
    });
  });

  it('Can generate OpenAPI using petstore', async () => {
    // Arrange
    const opts: Options = {
      openapi: 'tests/unit/petstore/petstore.yaml',
      output: testDirs[1]
    };

    // Act
    await main(opts);

    // Assert
    expect(mocks.write).toMatchSnapshot();
  });
});
