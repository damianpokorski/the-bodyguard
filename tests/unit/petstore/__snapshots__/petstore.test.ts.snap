// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Petstore generation tests Can generate OpenAPI using petstore 1`] = `
[MockFunction] {
  "calls": [
    [
      "Checking whether the required NPM packages are installed...
",
    ],
    [
      "  [92mâœ”[39m esbuild
",
    ],
    [
      "  [92mâœ”[39m @openapitools/openapi-generator-cli
",
    ],
    [
      "  [92mâœ”[39m typescript
",
    ],
    [
      "  [92mâœ”[39m corepack
",
    ],
    [
      "
Starting...
",
    ],
    [
      "Checking whether the required NPM packages are installed...
",
    ],
    [
      "  [92mâœ”[39m esbuild
",
    ],
    [
      "  [92mâœ”[39m @openapitools/openapi-generator-cli
",
    ],
    [
      "  [92mâœ”[39m typescript
",
    ],
    [
      "  [92mâœ”[39m corepack
",
    ],
    [
      "
Starting...
",
    ],
    [
      "  [92mâœ”[39m OpenAPI Specification validated
",
    ],
    [
      "
Extracting JSON Schemas...
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/schemas/order.json
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/schemas/category.json
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/schemas/user.json
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/schemas/tag.json
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/schemas/pet.json
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/schemas/apiResponse.json
",
    ],
    [
      "
Generating Typescript models...
",
    ],
    [
      "  - OpenAPI Generator CLI - (Typescript Axios)",
    ],
    [
      "  [92mâœ”[39m Models generated successfully using - openapi-generator-cli
",
    ],
    [
      "  - Ensuring camel case naming convention",
    ],
    [
      "  - Updating imports & Saving",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/models/apiResponse.ts
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/models/category.ts
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/models/index.ts
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/models/order.ts
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/models/pet.ts
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/models/tag.ts
",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/models/user.ts
",
    ],
    [
      "
Generating AJV Standalone validators (with better-ajv-errors)...
",
    ],
    [
      "  [92mâœ”[39m apiResponse - Generated... ",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/validators/apiResponse.ts
",
    ],
    [
      "  [92mâœ”[39m category - Generated... ",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/validators/category.ts
",
    ],
    [
      "  [92mâœ”[39m order - Generated... ",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/validators/order.ts
",
    ],
    [
      "  [92mâœ”[39m pet - Generated... ",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/validators/pet.ts
",
    ],
    [
      "  [92mâœ”[39m tag - Generated... ",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/validators/tag.ts
",
    ],
    [
      "  [92mâœ”[39m user - Generated... ",
    ],
    [
      "  [92mâœ”[39m /nas/projects/harness/tests/.builds.sample-generation/validators/user.ts
",
    ],
    [
      "
Compiling the final bundle...
",
    ],
    [
      "  [92mâœ”[39m Prepared the typescript interfaces
",
    ],
    [
      "  [92mâœ”[39m Mapped types for apiResponse
",
    ],
    [
      "  [92mâœ”[39m Mapped types for category
",
    ],
    [
      "  [92mâœ”[39m Mapped types for order
",
    ],
    [
      "  [92mâœ”[39m Mapped types for pet
",
    ],
    [
      "  [92mâœ”[39m Mapped types for tag
",
    ],
    [
      "  [92mâœ”[39m Mapped types for user
",
    ],
    [
      "Generating typescript declarion files before final bundling...",
    ],
    [
      "  [92mâœ”[39m Validated & generated declaration files using tsc
",
    ],
    [
      "Bundling & minifying... (esbuild / commonjs)",
    ],
    [
      "  [92mâœ”[39m Bundled and minified (esbuild)
",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
